lib/shoryuken/processor.rb

```diff
@@ -2,8 +2,75 @@ module Shoryuken
   class Processor
     include Util
 
-    def initialize(manager)
-      @manager = manager
+    BATCH_LIMIT = 10
+
+    def initialize(fetcher, polling_strategy)
+      @fetcher = fetcher
+      @polling_strategy = polling_strategy
+      # what is the worst thing that could happen because it's not atomic?
+      # is it possible that it evaluates to true even if we never set it to true?
+      @done = false
+    end
+
+    def start
+      # does updating and evaluating @thread has to be wrapped in mutex?
+      @thread ||= spawn_thread(&method(:loop))
+    end
+
+    def terminate(wait=false)
+      @done = true
+      # does updating and evaluating @thread has to be wrapped in mutex?
+      return if !@thread
+      @thread.value if wait
+    end
+
+    def kill
+      @done = true
+      # does updating and evaluating @thread has to be wrapped in mutex?
+      return if !@thread
+      @thread.raise ::Shoryuken::Shutdown
+    end
+
+    def alive?
+      # is this ok?
+      !!@thread.status
+    end
+
+    private
+
+    def loop
+      while !@done
+        do_work
+      end
+    rescue ::Shoryuken::Shutdown
+      # killed, bail!
+    rescue => e
+      # are we safe to retry here? What are kind of errors we do not want to recover from?
+      # OOM?
+      logger.debug { "Processor caught exception: #{e}" }
+      retry
+    end
+
+    def do_work
+      queue = @polling_strategy.next_queue
+      return logger.info { 'Doing nothing, all queues are paused' } if queue.nil?
+      batched_queue?(queue) ? dispatch_batch(queue) : dispatch_single_messages(queue)
+    end
+
+    def dispatch_batch(queue)
+      batch = @fetcher.fetch(queue, BATCH_LIMIT)
+      @polling_strategy.messages_found(queue.name, batch.size)
+      process(queue.name, patch_batch!(batch))
+    end
+
+    def dispatch_single_messages(queue)
+      messages = @fetcher.fetch(queue, 1)
+      @polling_strategy.messages_found(queue.name, messages.size)
+      messages.each { |message| process(queue.name, message) }
```

Comment:
@mariokostelac not sure if I'm following it, but you can't process all received messages using a single thread (one at time).  It would consume the visibility timeout from other messages, while the messages are idle waiting for the processor to process them.